### ISSUE TITLE
[pickers] Support UTC dates and timezones for DateFns

### ISSUE THREAD

**Issue by vu-dao-93** [mui/mui-x#10573]
### Search keywords

timezones date-fns datefns UTC

### Latest version

- [X] I have tested the latest version

### Summary 💡

We use `date-fns` as our date library, and just realized that the `timezone` prop only supports DayJs, Luxon & Moment. We have tried some workarounds, but the efforts seem like re-inventing the component. Could you make the `timezone` prop works for DateFns? What are the steps & challenges to make that happens?

### Examples 🌈

    <LocalizationProvider dateAdapter={AdapterDateFns}>
      <DateTimePicker timezone="UTC" />
    </LocalizationProvider>

**Comment by alexfauquette** [comment]
I kind of remind that the `date-fns` has not been added because of an issue. But I don't remember which one

@flaviendelangle is their a specific reason why date-fns has not be done or is it a lack of time?

**Comment by flaviendelangle** [comment]
Sorry for the delay, I missed the notification

Date FNS is using JavaScript `Date` object and you can't store a timezone on it (see [this article](https://bobbyhadz.com/blog/javascript-initialize-date-with-timezone))

THere is a version of `date-fns` called `date-fns-tz` which adds support for timezone ([NPM page](https://www.npmjs.com/package/date-fns-tz)).
But we need to see how it behaves exactly, because for me we won't be able to do things like `adapter.getTimezone`.

Basically it's just a question of priority for exploring what's doable.
But we have to keep in mind that we might hit a wall here.

**Comment by hutsul** [comment]
Hey there! Any news regarding this issue?

**Comment by LukasTy** [comment]
> Hey there! Any news regarding this issue?

I appreciate your interest. 🙏 
We are spread a bit too thin and have other priorities, so I can only confirm that we'll not be able to work on this issue before Q3. 😉 
I've set the grooming priority for this issue to "Medium" so as not to lose it in the depths of other issues.

**Comment by flaviendelangle** [comment]
`date-fns@4` has new APIs for timezone support, this should make this issue doable: https://blog.date-fns.org/v40-with-time-zone-support/

I'm moving it back to "To grooming to discuss the priority

**Comment by daniatic** [comment]
@flaviendelangle  This is crucial for our application. Can you tell us an ETA on this feature?

**Comment by comxd** [comment]
@flaviendelangle were you able to raise this issue during your discussions on priorities?

**Comment by calderon** [comment]
```
    "@mui/material": "5.16.14",
    "@mui/x-date-pickers": "7.27.3",
```

I have managed to bypass this situation, so I'll post the code here to help anyone in a similar situation. My project uses TS, and we utilise custom linter rules, such as `value == null`, to check if a value does not exist.

The idea is to have a time-zoned default value passed by prop and to one time-zoned state value to be used on the input.

I had another issue, but I'm not sure if it's related to the matter at hand. But if my initial value was `null` and I clicked on the selected date, the date input was never changed because the `handleDateChange` was never fired.

```ts
  const [inputValue, setInputValue] = useState(
    value == null
      ? ''
      : dateHelper.style({
        date: value,
        style
      })
  );
```

In my case, that date helper applies the timezone internally and "translates" the date to the desired format. Later on, on the component:

```ts
  const handleDateChange = (date: Date | null) => {
    onChange?.(date);
  };

return (
 <YourPickerComponent
   value={value ?? defaultCalendarValue}
   onAccept={handleDateChange}
   onChange={handleDateChange}
   timezone={timezone}
   format={format}
   slotProps={{
       day: {
          onClick: (event: MouseEvent<HTMLButtonElement>) => {
            const target = event.currentTarget;
            const isSelected = target.classList.contains('Mui-selected');

            if (value == null && isSelected) {
              onChange?.(dateHelper.tz(value));
            }
          }
        },
        actionBar: {
          onAccept: () => {
            if (value == null) {
              onChange?.(dateHelper.tz(value));
            }

            setOpen(false);
          },
        },
   }}
/>);
```

I hope this helps anyone with the same problem.

**Comment by chriswyatt** [comment]
This was my workaround:

```js
import { utc } from '@date-fns/utc';
import { addMinutes, subMinutes } from 'date-fns';

const tzOffset = op => dateObj => op(dateObj, dateObj.getTimezoneOffset());
const subTzOffset = tzOffset(subMinutes);
const addTzOffset = tzOffset(addMinutes);

// When setting DatePicker value
addTzOffset(local(v))

// When getting value from onChange() event
utc(subTzOffset(v))
```

The translation is done in a custom component I wrote, for use with React Hook Form. Because I'm in the UK and the app targets UK customers, it's easiest to just stick to UTC and only convert to local time, as and when needed. The hour difference is not important in most cases.

**Comment by afilp** [comment]
Will this be supported internally in MUI? We are also using `date-fns` with MUI.
